# Ejercicio 5
Alumna: Blanca Cano Camarero  
Correo: blancacano@correo.ugr.es

Dado un número $n$

In [1]:
n = 55883687

### Apartado primero 

Factoriza $n$ aplicando el método de la $\rho$ de Polard. ¿Cuántas iteraciones necesitas?
Sea $p1$ el mayor de sus factores primos y $p2$ el siguiente primo.

In [2]:
# Código auxiliar de prácticas ateriores
def Verbose_Rho_de_Polard(n:Integer, f) -> Integer:
    '''Devuelve un factor de p
    o n si es un fracaso
    '''
    mf = lambda x: f(x)% n
    #print(f"Cálculo de Rho de Polar para n={n}")
    x = 1
    y = 1
    d = 1
    paso = 0
    #print(f"Paso {paso}   x={x}    y={y}")
    while d == 1:
        paso += 1
        x = mf(x)
        y = mf(mf(y))
        d = gcd(abs(x-y),n)
        #print(f"Paso {paso}    x={x}    y={y}    mcd={d}")
    return d
f = lambda x : x^2 +1 

## Versión de sacar todos los factores de manera manual 
def SacaFactores(numero_factores, n, f ):
    factores = []
    m = n
    hay_factores = True 
    while numero_factores >= 0 and hay_factores:
        numero_factores -= 1
        d = Verbose_Rho_de_Polard(m,f)
        if d != m:
            factores.append(d)
            m = m // d
            print(f"Se tiene que n = multiplicar los elementos de {factores} y {m}\n")
        else:
            factores.append(m)
            print('='*60)
            hay_factores = False
    return factores

factores = SacaFactores(2, n,f )
print(f"Los factores de {n-1} son {factores}")

def Test_Lucas_Lehmer(n, bases):
    m = n-1
    for a in bases:
        # cumple condición 1
        if mod(a^m, n) == 1:
            condicion_dos = True
            for i in range (2,m):
                # si es divisor 
                if m % i == 0:
                    if (int(a**(m/i))% n) == 1:
                        condicion_dos =  False
                        break
            if condicion_dos == True:
                return a
    return 0

for p in factores:
    resultado = Test_Lucas_Lehmer(p,[2,3,5,7,11,13])
    mensaje = (
        f"p={p} es primo por el test Lucas-Lehmer con a = {resultado}",
        "No se ha encontrado un entero que satisfaga el test de Lucas-Lehmer para p = {p}"
    )[resultado == 0] 
    print(mensaje)

Se tiene que n = multiplicar los elementos de [7559] y 7393

Los factores de 55883686 son [7559, 7393]
p=7559 es primo por el test Lucas-Lehmer con a = 13
p=7393 es primo por el test Lucas-Lehmer con a = 5


A la vista de los resultado, tenemos que 
$p1 = 7559$ y $p2=7393$
### Apartado segundo   

Calcula las partes enteras de $\sqrt{p1}$ y $\sqrt{p2}$ con el algoritmo entero.

In [3]:
def VerboseParteEnteraRaizCuadrada(n):
    # Incicializamos a 
    a = n
    if is_even(n):
        a //= 2
    else:
        a = (a+1)//2
    menor = True 
    i = 1
    while menor:
        numerador = a**2 + n
        cociente =  numerador//(2*a)
        resto = numerador % (2*a)
        print(f"i:{i}  a_i:{a}  a^2 +n: {numerador} = {cociente} * {2*a} + {resto}")
        i += 1
        if cociente < a:
            a = cociente
        else:
            menor = False 
    print( "\n",f"La parte entera de  sqr({n}) es {a}", '\n','='*50,'\n')   
    return a 

for p in factores:
    VerboseParteEnteraRaizCuadrada(p)


def ParteEnteraRaizCuadrada(n):
    # Incicializamos a 
    a = n
    if is_even(n):
        a //= 2
    else:
        a = (a+1)//2
    menor = True 
    i = 1
    while menor:
        numerador = a**2 + n
        cociente =  numerador//(2*a)
        resto = numerador % (2*a)
        i += 1
        if cociente < a:
            a = cociente
        else:
            menor = False 
    return a 
    

i:1  a_i:3780  a^2 +n: 14295959 = 1890 * 7560 + 7559
i:2  a_i:1890  a^2 +n: 3579659 = 946 * 3780 + 3779
i:3  a_i:946  a^2 +n: 902475 = 476 * 1892 + 1883
i:4  a_i:476  a^2 +n: 234135 = 245 * 952 + 895
i:5  a_i:245  a^2 +n: 67584 = 137 * 490 + 454
i:6  a_i:137  a^2 +n: 26328 = 96 * 274 + 24
i:7  a_i:96  a^2 +n: 16775 = 87 * 192 + 71
i:8  a_i:87  a^2 +n: 15128 = 86 * 174 + 164
i:9  a_i:86  a^2 +n: 14955 = 86 * 172 + 163

 La parte entera de  sqr(7559) es 86 

i:1  a_i:3697  a^2 +n: 13675202 = 1849 * 7394 + 3696
i:2  a_i:1849  a^2 +n: 3426194 = 926 * 3698 + 1846
i:3  a_i:926  a^2 +n: 864869 = 466 * 1852 + 1837
i:4  a_i:466  a^2 +n: 224549 = 240 * 932 + 869
i:5  a_i:240  a^2 +n: 64993 = 135 * 480 + 193
i:6  a_i:135  a^2 +n: 25618 = 94 * 270 + 238
i:7  a_i:94  a^2 +n: 16229 = 86 * 188 + 61
i:8  a_i:86  a^2 +n: 14789 = 85 * 172 + 169
i:9  a_i:85  a^2 +n: 14618 = 85 * 170 + 168

 La parte entera de  sqr(7393) es 85 



### Apartado tercero 
Calcula la FCS de $\sqrt{p_1}$ y $\sqrt{p_2}$ con el algoritmo entero.

In [4]:
# Algoritmo 

def FraccionContinuaRaiz(p):
    parte_entera_raiz =  ParteEnteraRaizCuadrada(p)
    q_a = parte_entera_raiz
    FCS = [q_a]
    P_a = 0
    Q_a = 1
     
    for i in range(2*p-1):
        P_s = q_a*Q_a -P_a
        Q_s = (p - P_s*P_s)//Q_a
        q_s = (P_s + parte_entera_raiz)//Q_s
        FCS.append(q_s)
        P_a = P_s
        q_a = q_s
        Q_a = Q_s
    return FCS

def calcularPeriodo(ristra):
    ''' con esto calcuaremos el periodo de cada FCS
    Se fundamenta en que sabemos que son simétricos y que 
    acaba en el doble que el primero.
    '''
    primero = ristra[0]
    ultimo = 2*primero
    encontrado = False
    start = 1
    while not encontrado and start < len(ristra):
        start = ristra.index(ultimo,start)
        # Condición de simetría
        if ristra[1:start] == (ristra[1:start])[::-1]:
            encontrado = True 
    return start

numeros_que_mostrar = 200
for p in factores:
    FCS = FraccionContinuaRaiz(p)
    periodo = calcularPeriodo(FCS)
    print(f"La FCS de {p}, con periodo de tamaño {periodo} en los últimos elementos es:\n{FCS[:periodo+1]}\n")


La FCS de 7559, con periodo de tamaño 40 en los últimos elementos es:
[86, 1, 16, 2, 1, 1, 6, 2, 1, 3, 1, 8, 2, 1, 2, 1, 3, 1, 33, 1, 85, 1, 33, 1, 3, 1, 2, 1, 2, 8, 1, 3, 1, 2, 6, 1, 1, 2, 16, 1, 172]

La FCS de 7393, con periodo de tamaño 117 en los últimos elementos es:
[85, 1, 56, 3, 18, 1, 3, 2, 5, 1, 12, 2, 1, 1, 1, 1, 3, 1, 3, 1, 6, 2, 1, 2, 21, 8, 7, 24, 2, 2, 1, 7, 2, 9, 1, 1, 1, 4, 1, 2, 1, 1, 4, 1, 1, 1, 2, 1, 14, 1, 9, 1, 4, 3, 3, 3, 1, 2, 3, 3, 2, 1, 3, 3, 3, 4, 1, 9, 1, 14, 1, 2, 1, 1, 1, 4, 1, 1, 2, 1, 4, 1, 1, 1, 9, 2, 7, 1, 2, 2, 24, 7, 8, 21, 2, 1, 2, 6, 1, 3, 1, 3, 1, 1, 1, 1, 2, 12, 1, 5, 2, 3, 1, 18, 3, 56, 1, 170]



Además sabemos que quitando el primer número de la tupla presentada el resto es el periodo. 

# Ejercicio 6
Alumna: Blanca Cano Camarero  
Correo: blancacano@correo.ugr.es


Sea $p$ el factor del ejercicio 5 con mayor periodo, en mi caso $p=7393.$

In [5]:
p = factores[1]
print(f"p = {p}")

FCS_p = FraccionContinuaRaiz(p)
periodo = calcularPeriodo(FCS)
FCS_p = FCS_p[:periodo+1]
print(FCS_p)

p = 7393
[85, 1, 56, 3, 18, 1, 3, 2, 5, 1, 12, 2, 1, 1, 1, 1, 3, 1, 3, 1, 6, 2, 1, 2, 21, 8, 7, 24, 2, 2, 1, 7, 2, 9, 1, 1, 1, 4, 1, 2, 1, 1, 4, 1, 1, 1, 2, 1, 14, 1, 9, 1, 4, 3, 3, 3, 1, 2, 3, 3, 2, 1, 3, 3, 3, 4, 1, 9, 1, 14, 1, 2, 1, 1, 1, 4, 1, 1, 2, 1, 4, 1, 1, 1, 9, 2, 7, 1, 2, 2, 24, 7, 8, 21, 2, 1, 2, 6, 1, 3, 1, 3, 1, 1, 1, 1, 2, 12, 1, 5, 2, 3, 1, 18, 3, 56, 1, 170]


### Apartado primero 
Calcula los convergentes de $\sqrt p$

In [11]:
# Calculamos los convergentes 

def Convergentes(FCS):
    A_aa = 1
    A_a = FCS[0]
    B_aa = 0
    B_a = 1
    convergentes = [A_a/B_a]
    convergentes_A = [A_a]
    convergentes_B = [B_a]
    for q in FCS[1:]:
        A_s = q* A_a + A_aa
        B_s = q*B_a + B_aa
        convergentes.append(A_s /B_s)
        convergentes_A.append(A_s)
        convergentes_B.append(B_s)
        #actualizamos pesos
        A_aa, A_a = A_a, A_s
        B_aa, B_a = B_a, B_s
    return convergentes, convergentes_A, convergentes_B

print(f"Los convergentes de sqrt({p}) son:")
convergentes, convergentes_A, convergentes_B = Convergentes(FCS_p)
for i in range(len(convergentes)):
    print(f"A_{i} =  {convergentes_A[i]}")
    print(f"B_{i} = {convergentes_B[i]}")


Los convergentes de sqrt(7393) son:
A_0 =  85
B_0 = 1
A_1 =  86
B_1 = 1
A_2 =  4901
B_2 = 57
A_3 =  14789
B_3 = 172
A_4 =  271103
B_4 = 3153
A_5 =  285892
B_5 = 3325
A_6 =  1128779
B_6 = 13128
A_7 =  2543450
B_7 = 29581
A_8 =  13846029
B_8 = 161033
A_9 =  16389479
B_9 = 190614
A_10 =  210519777
B_10 = 2448401
A_11 =  437429033
B_11 = 5087416
A_12 =  647948810
B_12 = 7535817
A_13 =  1085377843
B_13 = 12623233
A_14 =  1733326653
B_14 = 20159050
A_15 =  2818704496
B_15 = 32782283
A_16 =  10189440141
B_16 = 118505899
A_17 =  13008144637
B_17 = 151288182
A_18 =  49213874052
B_18 = 572370445
A_19 =  62222018689
B_19 = 723658627
A_20 =  422545986186
B_20 = 4914322207
A_21 =  907313991061
B_21 = 10552303041
A_22 =  1329859977247
B_22 = 15466625248
A_23 =  3567033945555
B_23 = 41485553537
A_24 =  76237572833902
B_24 = 886663249525
A_25 =  613467616616771
B_25 = 7134791549737
A_26 =  4370510889151299
B_26 = 50830204097684
A_27 =  105505728956247947
B_27 = 1227059689894153
A_28 =  215381968801647

### Apartado segundo 

Calcula las soluciones de las ecuaciones de Pell, $x^2 -py^2 = \pm 1$

Las soluciones positivas de $x^2 -py^2 = \pm 1$ son los convergentes de $A_{n-1}, B_{n-1}$ de $\sqrt{p}$ con $n$ múltiplo del la longitud del periodo. 

In [7]:
x = convergentes_A[-2]
y = convergentes_B[-2]

print(f"Comprobemos que para \nx={x} y \ny={y} \nx^2 - p*y^2 ={x^2 - p*y^2}")

print("Las soluciones de la ecuación de Pell son:")
index = 0
for i,j in [(1,1), (-1,1), (1,-1), (-1,-1)]:
    print(f"x_{index} = {x*i}")
    print(f"y_{index} = {y*j}", "\n")
    index += 1


Comprobemos que para 
x=1065054648479261600011588720503279995290036819652322930864 y 
y=12386868842214623369411176602981866205578633359304869777 
x^2 - p*y^2 =-1
Las soluciones de la ecuación de Pell son:
x_0 = 1065054648479261600011588720503279995290036819652322930864
y_0 = 12386868842214623369411176602981866205578633359304869777 

x_1 = -1065054648479261600011588720503279995290036819652322930864
y_1 = 12386868842214623369411176602981866205578633359304869777 

x_2 = 1065054648479261600011588720503279995290036819652322930864
y_2 = -12386868842214623369411176602981866205578633359304869777 

x_3 = -1065054648479261600011588720503279995290036819652322930864
y_3 = -12386868842214623369411176602981866205578633359304869777 



Observemos que es natural que $x^2 - py^2 = -1$ pues el periodo $r = 117$ es impar. 
Las demás soluciones positivas $(x_n, y_n)$ se obtiene de la expresión

$$x_n + y_n \sqrt{p} = (A_{r-1} + B _{r-1})^n$$

con $n$ natural. Las parejas correspondientes a $n$ par son soluciones de $x^2 - py^2 = 1$ y para $n$ impar lo es de $x^2 - py^2 = -1.$

### Apartado tercero

Calcula las unidades del anillo de enteros cuadráticos $Z[\sqrt{p}]$
---

Las unidades de $Z[\sqrt{p}]$  son exactamente los elementos 
$x + y \sqrt p$ con norma $\pm 1$, es decir los elementos que satisfacen una de las ecuaciones de Pell. 

Es decir que tomando los $x_n$ $y_n$ solución del apartado anterior son los elementos de la forma 
$x_n + y_n \sqrt p$, $-x_n + y_n \sqrt p$, $x_n - y_n \sqrt p$ y 
$-x_n - y_n \sqrt p.$


### Apartado cuarto 

¿Es $Z[\sqrt{p}]$ el anillo de enteros del cuerpo $Q[\sqrt{p}]$?

Se tiene que en nuestro caso 


In [9]:
mod(p,4)

1

Luego el anillo de entero de $Q[\sqrt{p}]$ es 

$Z[ \frac{1+\sqrt{p}}{2}]$


Puesto que  $Z[\sqrt{p}] \subsetneq Z[ \frac{1+\sqrt{p}}{2}]$